include/boost/math/differentiation/autodiff.hpp

```diff
@@ -0,0 +1,1882 @@
+    for (size_t i=1 ; i<=order_sum() ; ++i)
+    {    // accumulator += (epsilon_i *= epsilon) * f(i);
+        epsilon_i = epsilon_i.epsilon_multiply(i-1, 0, epsilon, 1, 0);
+        accumulator += epsilon_i.epsilon_multiply(i, 0, f(i));
+    }
+    return accumulator;
+}
+
+// f : order -> derivative(order)
+template<typename RealType,size_t Order>
+dimension<RealType,Order> dimension<RealType,Order>::apply_with_horner(const std::function<root_type(size_t)>& f) const
+{
+    const dimension<RealType,Order> epsilon = dimension<RealType,Order>(*this).set_root(0);
+    auto accumulator = dimension<RealType,Order>{f(order_sum())/boost::math::factorial<root_type>(order_sum())};
+    for (size_t i=order_sum() ; i-- ;)
+        (accumulator *= epsilon) += f(i) / boost::math::factorial<root_type>(i);
+    return accumulator;
+}
+
+// f : order -> derivative(order)/factorial(order)
+// Use this when the computation of the derivatives already includes the factorial terms. E.g. See atan().
+template<typename RealType,size_t Order>
+dimension<RealType,Order>
+    dimension<RealType,Order>::apply_with_horner_factorials(const std::function<root_type(size_t)>& f) const
+{
+    const dimension<RealType,Order> epsilon = dimension<RealType,Order>(*this).set_root(0);
+    auto accumulator = dimension<RealType,Order>{f(order_sum())};
+    for (size_t i=order_sum() ; i-- ;)
+        (accumulator *= epsilon) += f(i);
+    return accumulator;
+}
+
+#ifndef BOOST_NO_CXX17_IF_CONSTEXPR
+// Can throw "std::out_of_range: array::at: __n (which is 7) >= _Nm (which is 7)"
+template<typename RealType,size_t Order>
+template<typename... Orders>
+typename type_at<RealType,sizeof...(Orders)>::type dimension<RealType,Order>::at(size_t order, Orders... orders) const
+{
+    if constexpr (0 < sizeof...(orders))
+        return v.at(order).at(orders...);
+    else
+        return v.at(order);
+}
+#endif
+
+#ifndef BOOST_NO_CXX17_IF_CONSTEXPR
+template<typename RealType,size_t Order>
+constexpr size_t dimension<RealType,Order>::depth()
+{
+    if constexpr (is_dimension<RealType>::value)
+        return 1 + RealType::depth();
+    else
+        return 1;
+}
+#endif
+
+#ifndef BOOST_NO_CXX17_IF_CONSTEXPR
+template<typename RealType,size_t Order>
+constexpr size_t dimension<RealType,Order>::order_sum()
+{
+    if constexpr (is_dimension<RealType>::value)
+        return Order + RealType::order_sum();
+    else
+        return Order;
+}
+#endif
+
+#ifndef BOOST_NO_CXX17_IF_CONSTEXPR
+// Can throw "std::out_of_range: array::at: __n (which is 7) >= _Nm (which is 7)"
+template<typename RealType,size_t Order>
+template<typename... Orders>
+typename type_at<RealType,sizeof...(Orders)-1>::type dimension<RealType,Order>::derivative(Orders... orders) const
+{
+    static_assert(sizeof...(orders) <= depth(),
+        "Number of parameters to derivative(...) cannot exceed the number of dimensions in the dimension<...>.");
+    return at(orders...) * (... * boost::math::factorial<root_type>(ord
```

Comment:
Is it idiomatic to set the precision on the ostream in a call? Presumably people want to set their own precision.

Reply:
Agreed; setting the precision has been removed.